### DISCUSSION TITLE
Use python typing system for composable apps

### DISCUSSION THREAD

**Discussion by GavinHuttley** [cogent3/cogent3#1071] [Ideas]
At present I define input/output data types for `cogent3` composable apps in a very primitive way -- [as strings]((https://github.com/cogent3/cogent3/blob/48b4d85628c731c7af5771dd6625838402d7d885/src/cogent3/app/composable.py#L125)).

Should this approach be changed to using python's type systems?

For instance, if we did the following
```python
from typing import TypeVar

AlignmentType = TypeVar("AlignmentType")
AnotherType = TypeVar("AnotherType")
```
This approach would require we change the definitions of composable apps and how the composability of two apps is evaluated.

For example, currently we have the following style of class

```python
class A(Composable):
    _input_types =(AlignmentType,)
    _output_types = (AlignmentType,)

    def __init__(self):
        super(A, self).__init__(
            input_types=self._input_types,
            output_types=self._output_types,
        )
        self._in = None
        self._out = None
```

What if we change the Composable classes to use annotations, e.g.
```python
class A(Composable):
    _in: AlignmentType
    _out: AlignmentType
```

If we then define 2 other apps
```python
class B(Composable):
    _in: AnotherType
    _out: AnotherType

class B(Composable):
    _in: AlignmentType
    _out: AnotherType
```

How would we check composability such that `A() + B()` fails but `A() + C()` works?

**Comment by GavinHuttley** [comment]
@Nick-Foto Can you evaluate whether the type system can be used in this way? I suggest not using `cogent3` at all here, rather just writing a couple of demo classes that explore this.

**Reply by GavinHuttley** [reply]
An important point to make here is that we will always be using custom `cogent3` types or `Union` of such types, so much of what these type checking libraries provide is overkill.

Additionally, this is not a performance critical part of the code so speed is not an issue.

**Reply by Nick-Foto** [reply]
I did some research about it. Python typing/PEP 484 type system is built for static type checking and is inappropriate for dynamic type checking.
These type annotations can be used statically by third-party packages like `mypy` to check typing, not runtime checking and failure for type mismatch, like the above example A() + B() fails but A() + C() works.

**Comment by GavinHuttley** [comment]
# Accessing type hints

```python
from typing import Union, TypeVar, get_args, get_type_hints

AlignmentType = TypeVar("AlignmentType")
AnotherType = TypeVar("AnotherType")
U = Union[AlignmentType, AnotherType]

class MyClass:
    x: U

m = MyClass()

hints = get_type_hints(m)
AlignmentType in get_args(hints["x"])  # returns True
```

The answer to the question can be built using the above. One issue is I think whether this approach will work in python 3.7?

**Comment by GavinHuttley** [comment]
## Creating type hints at runtime

```python
union = Union[(A_T, B_T)]  # note the mixture must be `tuple`
```
## Assigning type hints at runtime

Reading the [python docs](https://docs.python.org/3/howto/annotations.html#best-practices-for-annotations-in-any-python-version), this requires directly modifying the `__annotations__` dict. Knowing there's an attribute `x` on the class, we do
```python
m.__annotations__["x"] = union
```

**Reply by Nick-Foto** [reply]
According to ["Annotations Best Practices"](https://docs.python.org/3.11/howto/annotations.html#annotations-best-practices), we should avoid modifying `__annotations__` dict directly.  In that article explained how we can access and modify data inside `__annotations__` dict.
Python 3.10 adds a new function to the standard library: [inspect.get_annotations()](https://docs.python.org/3.11/library/inspect.html#inspect.get_annotations). In Python versions 3.10 and newer, calling this function is the best practice for accessing the annotations dict of any object that supports annotations.
In Python 3.9 and older, we should use three-argument [getattr()](https://docs.python.org/3.11/library/functions.html#getattr) to access the object’s `__annotations__` attribute, because some  classes may not have `__annotations__` dict.
Only in Python 3.10 and above, only for functions, classes and modules, we can safely use `object.__annoations__`. for other objects, or unknown objects, we should use `getattr` function.
And finally, I am not sure annotations at runtime can be useful, because all of these typing system, as far as I know, is for static type checking only.

**Comment by GavinHuttley** [comment]
The advantage of using type hints are:
- we adopt a standard approach that developers will know
- code can be validated using type checkers
- the code expression can be more succinct, i.e. instead of providing a tuple of enumerated names we can just do `Union[type1,type2]` or (in python 3.10+) `type1 | type2`.
- these can be set directly on the attributes of composable subclasses and will not require passing arguments to super class constructor

So, answering my own question, I think we should use these type hints instead of the custom string constants.

**Reply by Nick-Foto** [reply]
I found a package for type-checking at rune-time. [strongtyping](https://strongtyping.readthedocs.io/en/latest/#the-solution)

**Reply by Nick-Foto** [reply]
I also found two other packages for type-checking at rune-time. 
[Pyright](https://github.com/microsoft/pyright)
[Typeguard](https://typeguard.readthedocs.io/en/latest/#typeguard)

**Reply by GavinHuttley** [reply]
We always tradeoff adding a dependency versus the complexity of our requirement.
At this point I don't think a dependency will work because we have to provide some backwards compatibility for the existing approach.
I have created a new git branch `composable_refactor` where we can experiment with my proposed approach.

**Reply by GavinHuttley** [reply]
But for completeness, here's another run-time type checking package
- [beartype](https://pypi.org/project/beartype/)
